### ISSUE TITLE
check_parent performs a recursive search, but it shouldn't ?

### ISSUE THREAD

**Issue by roadelou** [theodumont/video-logging#2]
The tool moves and rearranges files from only the current directory, yet the check_parent function will look for the .videolog file in the current working directory _and all all of its sub-directories recursively_. This is what os.walk does in [src/video_logging.py](https://github.com/theodumont/video-logging/blob/master/src/video_logging.py) :

```python3
for root, dirs, files in os.walk("./"):
        if ".videolog" in files:
            raise SudoException()
```

I do not think that there is an equivalent to _os.listdir_ but for files in python. One way to circumvent this is to take advantage of the fact that os.walk returns a __generator__ object (and not a list). This means that instead of getting all the files recursively at once, os.walk (very probably) gets those in the current working directory, then those in the first sub-directory, etc...

Hence, one way to only get the files in the current working directory _without going through every sub-directory_ would be something like :

```python3
for root, dirs, files in os.walk("./"):
        if root == '.':
            # Do your stuff
            break
        # else continue
```

Since the top directory is likely the first on to be walked through, this shouldn't be too inefficient.

That should also solve the issue mentioned in [src/data.json](https://github.com/theodumont/video-logging/blob/master/src/data.json) in the __sudo__ section :

> [B]y default, the tool checks if the current directory contains the 'video-logging' folder in its arborescence by looking for the '.videolog' file. This process is usually quite fast but it may take a while when dealing with big directories.

On most modern file systems, I think that __file look-ups in a given directory__ should be of almost __constant time__ (correct me if I am wrong). Hence this issue is probably linked with the os.walk thing.

**Comment by theodumont** [comment]
Yes, I did want to check the current working directory and each of its subdirectories, and not only the `cwd`. Unless I didn't get your point? 🧐

Actually, as the `folder` command does not perform a recursive sort of the files and only moves the files that are in the current working directory, sorting a directory that contains `.videolog` could have 2 different consequences:
- if `cwd` _is_ `video-logging` directly, `cli.py` and `src/` would be put in separated folders and thus one __could no longer use the tool__ after the `folder` command;
- if `cwd` _contains_ the `video-logging` folder, then the whole folder will be moved and the tool __will still be functional__, but in another location on the computer.

So in a way, it may be useless to check all of the subdirectories, but I chose to do this so that the project folder doesn't move.
But I'm thinking about dropping the recursive check and only reading the content of the current directory; some people may want to sort their `repos` folder, which may contain this one.
I was thinking about using
```py
os.path.isfile(".videolog")
```
for this.


Anyway, what do you think about this method, _i.e._ having an empty file that is only used for checking if we are in a specific directory? Another idea - that could be worse - would be to keep in memory the path in which the script was executed, and to compare it with the `cwd` each time.

**Comment by roadelou** [comment]
Yes, we are talking about the same thing :smile:

I feel that __my initial issue lacked some important information__ and might as such have been confusing :dizzy:, allow me to correct that :wink:

In the _current_ implementation, you have chosen to __trade usability__ (the tool takes time to perform the os.walk _out of the box_) __for safety__ (the tool will not break itself _out of the box_) :balance_scale:

But, as you might have noticed, __not many tools do that__:exclamation:
 - Python lets you delete the python interpreter :snake:
 - Windows lets you delete registry keys :wink:

In fact, __almost no tool will check whether what you are doing makes sense__, so long as it __can perform what you requested__:
 - You can __open and edit a video with a text editor__ :pencil: There is little point in doing so (from a human perspective) but for as far as the editor is concerned __this is perfectly valid manipulation__ :heavy_check_mark:
 - Windows will let you move :arrow_right: its internal libraries (even if that will break the OS on reboot) ___because it can___.

There is no deep philosophical reason for that, it is a __limitation inherited from earlier days of computer science__ (see the C programming language) :laughing: This is sometimes (_rarely in fact_) referred to as a __best effort__ policy, i.e. the __tool will try to perform what the human requested to the fullest of its abilities__.

---

This is opposed to __dynamically deterministic__ tools that twill always produce a reliable result, but might restrain themselves. A python interpreter :snake: that would __only execute a script if it terminates__ would be an example of this: it __cannot execute all scripts__, but if it does run one then we have a __reliable guarantee__ that the execution will eventually end.

If you are interested in that sort of things I would advise you to investigate __functional programming__ with languages such as [Haskell](https://www.haskell.org/) or [Rust](https://www.rust-lang.org/).

---

__Historically__, the reason why most tools didn't do that sort of verification is because __it would have been to complex__ (or impossible). But the whole programming _environment evolved to account for this weakness_ :bulb:, and in particular the __tools__ :wrench: (applications, utilities) are __never stored along user data__ :book: in the computer.

For instance on Windows, __tools__ :wrench: can be found in __C:\\Program Files\\__ or __C:\\Program Files(x64)\\__ (for system-wide installs) or __C:\\Users\\\<username\>\\AppData\\\<some more things\>__ (for user-specific tools). On the other hand, __user data__ :book: would be found in __C:\\Users\\\<username\>\\Documents__ for instance.

Thus, my issue with this trade off is that you are __giving up on usability__ (slower on large directories _out of the box_) to account for an __edge case that should never happen__. Plus it is kind of surprising when you read it :smile:

With that being said, there is at least one notable exception to this: when you are __debugging the tool__, you will almost always do this _next to its source code_. Thus the problem you are tackling is a __real issue for development__, and I guess you wrote it because you had some misfortune in the first place :wink:

But as such I believe __this verification could be opt-in__, and not opt-out as it is now. So instead of asking the tool to remove this safeguard :cop: with "sudo", you could instead __add a debug flag__ that would raise it. That is just an example of how to do that, probably you will find a more suitable way.

About __how the verification should be performed__, I think that all the ways you have mentioned and tried are valid, although __I like the one where you dynamically find the path of the executable__ :heart: more, because it should have no false-positive :heavy_check_mark:
